arbnode/espresso_caff_node.go

```diff
@@ -0,0 +1,286 @@
+package arbnode
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	espressoClient "github.com/EspressoSystems/espresso-sequencer-go/client"
+	flag "github.com/spf13/pflag"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/offchainlabs/nitro/arbos"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/espressostreamer"
+	"github.com/offchainlabs/nitro/execution/gethexec"
+	"github.com/offchainlabs/nitro/solgen/go/bridgegen"
+	"github.com/offchainlabs/nitro/util/headerreader"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+type DelayedMessageFetcherInterface interface {
+	getDelayedMessage(index uint64) (*arbostypes.L1IncomingMessage, error)
+}
+
+type DelayedMessageFetcher struct {
+	fromBlock     uint64
+	delayedBridge *DelayedBridge
+	l1Reader      *ethclient.Client
+
+	delayedMessages map[uint64]*arbostypes.L1IncomingMessage
+}
+
+func NewDelayedMessageFetcher(delayedBridge *DelayedBridge, l1Reader *ethclient.Client) *DelayedMessageFetcher {
+	return &DelayedMessageFetcher{
+		fromBlock:       delayedBridge.fromBlock,
+		delayedBridge:   delayedBridge,
+		l1Reader:        l1Reader,
+		delayedMessages: make(map[uint64]*arbostypes.L1IncomingMessage),
+	}
+}
+
+func (f *DelayedMessageFetcher) getDelayedMessage(index uint64) (*arbostypes.L1IncomingMessage, error) {
+	if f.delayedMessages[index] != nil {
+		msg := f.delayedMessages[index]
+		delete(f.delayedMessages, index)
+		return msg, nil
+	}
+
+	currL1, err := f.l1Reader.BlockNumber(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	if currL1 < f.fromBlock {
+		return nil, fmt.Errorf("l1 block number %d is less than from block %d", currL1, f.fromBlock)
+	}
+
+	startBlock := f.fromBlock
+	endBlock := currL1
+	hasFound := false
+
+	// This value is from the InboxReader default blocks-to-read.
+	// TODO: Make this configurable.
+	blocksToRead := uint64(100)
+
+	for startBlock <= endBlock && !hasFound {
```

Comment:
We might be able to improve this later to lazily fetch the data by breaking when `hasfound = true` and the message is added to `f.delayedMessages`, but lets get the mvp done first.

Reply:
I am not sure what you meant. I think the current implementation is already lazy and it breaks when it finds the target delayed message

Reply:
Yeah you're right I missed the hasFound in the outer for loop. My bad.